source/common/http/websocket/ws_handler_impl.h

```diff
@@ -0,0 +1,93 @@
+#pragma once
+
+#include <string>
+
+#include "envoy/http/filter.h"
+#include "envoy/network/connection.h"
+#include "envoy/network/filter.h"
+#include "envoy/upstream/cluster_manager.h"
+#include "envoy/upstream/upstream.h"
+
+#include "common/common/logger.h"
+#include "common/network/filter_impl.h"
+
+namespace Envoy {
+namespace Http {
+namespace WebSocket {
+
+/**
+ * An implementation of a WebSocket proxy based on TCP proxy. This filter will instantiate a
+ * new outgoing TCP connection using the defined load balancing proxy for the configured cluster.
+ * All data will be proxied back and forth between the two connections, without any knowledge of
+ * the underlying WebSocket protocol.
+ *
+ * N.B. This class implements Network::ReadFilter interfaces purely for sake of consistency
+ * with TcpProxy filter. WsHandlerImpl it is not used as a network filter in any way.
+ */
+class WsHandlerImpl : public Network::ReadFilter, Logger::Loggable<Logger::Id::websocket> {
```

Comment:
That is a good point. My aim is to move most of ws_handler out, and reuse TCP. However, the crux of ws_handler is in initializeUpstreamConnection (where all this error handling is done in Websocket specific way - for example, no tcp backend, send http 503/404).

I think I had noted this elsewhere in the PR. If there is a way to latch onto a backend connection for a request (we would need this for sticky sessions anyway, if we plan to support it), then this entire WS implementation can reuse HTTP/1.1 code, and drop into tcp only when the websocket upgrade is complete.